Building with LLVM 6 fails

**Godot version:**
Build the latest zip snapshot of master.


**OS/device including version:**
Hardware:

- Ryzen 3900X
- Sapphire Pulse Radeon RX 580
- 128GB RAM

Software:

- elementary OS 5.1.7 Hera (based on Ubuntu 18.04.4 LTS)
- Linux 5.4.0-45-generic
- GTK 3.22.30
- Mesa 20.1.7kisak2b
- gcc/bionic-updates,bionic-security,now 4:7.4.0-1ubuntu2.3 amd64 [installed,automatic]
- llvm-10/bionic,now 1:10.0.1-1\~kisak1\~b amd64 [installed]
- clang-10/bionic,now 1:10.0.1-1\~kisak1\~b amd64 [installed]

**Issue description:**
_(This issue was encountered while trying build option variations to see if they affected the LLVM LTO problem in #40279)_

Building with LLVM fails with the following compile errors in `modules/gdscript/gdscript_cache.cpp`:
```
modules/gdscript/gdscript_cache.cpp:111:36: error: expected ')'
        MutexLock(GDScriptCache::singleton->lock);
                                          ^
modules/gdscript/gdscript_cache.cpp:111:11: note: to match this '('
        MutexLock(GDScriptCache::singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:111:27: error: definition or redeclaration of 'singleton' not allowed inside a function
        MutexLock(GDScriptCache::singleton->lock);
                  ~~~~~~~~~~~~~~~^
modules/gdscript/gdscript_cache.cpp:118:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:118:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:118:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:118:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
modules/gdscript/gdscript_cache.cpp:124:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:124:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:124:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:124:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
modules/gdscript/gdscript_cache.cpp:171:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:171:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:171:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:171:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
modules/gdscript/gdscript_cache.cpp:193:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:193:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:193:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:193:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
```


**Steps to reproduce:**

1. Download a snapshot of the Godot source and extract it.
2. Go into the project folder and build, making sure that the LLVM build option is enabled (other LLVM related options do not affect the result. Nor do the values given for the `-j` or `optimize` options).
**Example Build Command:**
`scons -j24 platform=x11 target=debug optimize=size use_llvm=yes use_lld=yes`
3. The build fails when compiling `modules/gdscript/gdscript_cache.cpp`.


**Minimal reproduction project:**
_(not applicable)_

No. Those are the only error messages, copied directly from the terminal window.

Here's the entire output of the latest build attempt:
```
$ > scons -j24 platform=x11 target=debug use_llvm=yes
scons: Reading SConscript files ...
Platform "x11" has been renamed to "linuxbsd" in Godot 4.0. Building for platform "linuxbsd".
Enabling ALSA
Enabling PulseAudio
Checking for C header file mntent.h... (cached) yes
scons: done reading SConscript files.
scons: Building targets ...
[ 31%] Compiling ==> modules/gdscript/gdscript_cache.cpp
[ 33%] Linking Static Library ==> modules/libmodule_gdnavigation.linuxbsd.tools.64.llvm.a
[ 33%] Ranlib Library         ==> modules/libmodule_gdnavigation.linuxbsd.tools.64.llvm.a
[ 33%] Compiling ==> modules/gdnative/android/android_gdn.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/aabb.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/array.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/basis.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/callable.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/color.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/dictionary.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/gdnative.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/node_path.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/packed_arrays.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/plane.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/quat.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/rect2.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/rid.cpp
[ 33%] Compiling ==> modules/gdnative/gdnative/string.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative/string_name.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative/transform.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative/transform2d.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative/variant.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative/vector2.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative/vector3.cpp
[ 34%] Compiling ==> modules/gdnative/nativescript/api_generator.cpp
[ 34%] Compiling ==> modules/gdnative/nativescript/godot_nativescript.cpp
[ 34%] Compiling ==> modules/gdnative/nativescript/nativescript.cpp
[ 34%] Compiling ==> modules/gdnative/nativescript/register_types.cpp
modules/gdscript/gdscript_cache.cpp:111:36: error: expected ')'
        MutexLock(GDScriptCache::singleton->lock);
                                          ^
modules/gdscript/gdscript_cache.cpp:111:11: note: to match this '('
        MutexLock(GDScriptCache::singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:111:27: error: definition or redeclaration of 'singleton' not allowed inside a function
        MutexLock(GDScriptCache::singleton->lock);
                  ~~~~~~~~~~~~~~~^
modules/gdscript/gdscript_cache.cpp:118:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:118:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:118:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:118:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
modules/gdscript/gdscript_cache.cpp:124:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:124:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:124:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:124:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
modules/gdscript/gdscript_cache.cpp:171:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:171:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:171:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:171:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
modules/gdscript/gdscript_cache.cpp:193:21: error: expected ')'
        MutexLock(singleton->lock);
                           ^
modules/gdscript/gdscript_cache.cpp:193:11: note: to match this '('
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:193:11: error: cannot use parentheses when declaring variable with deduced class template specialization type
        MutexLock(singleton->lock);
                 ^
modules/gdscript/gdscript_cache.cpp:193:12: error: declaration of variable 'singleton' with deduced type 'MutexLock' requires an initializer
        MutexLock(singleton->lock);
                  ^
[ 34%] Compiling ==> modules/gdnative/gdnative_library_singleton_editor.cpp
[ 34%] Compiling ==> modules/gdnative/gdnative_library_editor_plugin.cpp
[ 34%] Compiling ==> modules/gdnative/net/multiplayer_peer_gdnative.cpp
[ 34%] Compiling ==> modules/gdnative/net/packet_peer_gdnative.cpp
[ 34%] Compiling ==> modules/gdnative/net/register_types.cpp
[ 34%] Compiling ==> modules/gdnative/net/stream_peer_gdnative.cpp
[ 34%] Compiling ==> modules/gdnative/net/webrtc_gdnative.cpp
[ 34%] Compiling ==> modules/gdnative/xr/register_types.cpp
[ 34%] 14 errors generated.
Compiling ==> modules/gdnative/xr/xr_interface_gdnative.cpp
scons: *** [modules/gdscript/gdscript_cache.linuxbsd.tools.64.llvm.o] Error 1
modules/gdnative/xr/xr_interface_gdnative.cpp:273:2: warning: this needs to be redone [-W#warnings]
#warning this needs to be redone
 ^
modules/gdnative/xr/xr_interface_gdnative.cpp:290:2: warning: need to obtain this ID again [-W#warnings]
#warning need to obtain this ID again
 ^
2 warnings generated.
modules/gdnative/nativescript/nativescript.cpp:116:2: warning: inheritance needs to be implemented in NativeScript [-W#warnings]
#warning inheritance needs to be implemented in NativeScript
 ^
1 warning generated.
scons: building terminated because of errors.
```

--- a/modules/gdscript/gdscript_cache.cpp
+++ b/modules/gdscript/gdscript_cache.cpp
@@ -108,20 +108,20 @@ GDScriptParserRef::~GDScriptParserRef() {
 	if (analyzer != nullptr) {
 		memdelete(analyzer);
 	}
-	MutexLock(GDScriptCache::singleton->lock);
+	MutexLock lock(GDScriptCache::singleton->lock);
 	GDScriptCache::singleton->parser_map.erase(path);
 }
 
 GDScriptCache *GDScriptCache::singleton = nullptr;
 
 void GDScriptCache::remove_script(const String &p_path) {
-	MutexLock(singleton->lock);
+	MutexLock lock(singleton->lock);
 	singleton->shallow_gdscript_cache.erase(p_path);
 	singleton->full_gdscript_cache.erase(p_path);
 }
 
 Ref<GDScriptParserRef> GDScriptCache::get_parser(const String &p_path, GDScriptParserRef::Status p_status, Error &r_error, const String &p_owner) {
-	MutexLock(singleton->lock);
+	MutexLock lock(singleton->lock);
 	Ref<GDScriptParserRef> ref;
 	if (p_owner != String()) {
 		singleton->dependencies[p_owner].insert(p_path);
@@ -168,7 +168,7 @@ String GDScriptCache::get_source_code(const String &p_path) {
 }
 
 Ref<GDScript> GDScriptCache::get_shallow_script(const String &p_path, const String &p_owner) {
-	MutexLock(singleton->lock);
+	MutexLock lock(singleton->lock);
 	if (p_owner != String()) {
 		singleton->dependencies[p_owner].insert(p_path);
 	}
@@ -190,7 +190,7 @@ Ref<GDScript> GDScriptCache::get_shallow_script(const String &p_path, const Stri
 }
 
 Ref<GDScript> GDScriptCache::get_full_script(const String &p_path, Error &r_error, const String &p_owner) {
-	MutexLock(singleton->lock);
+	MutexLock lock(singleton->lock);
 
 	if (p_owner != String()) {
 		singleton->dependencies[p_owner].insert(p_path);

